Fix third party cookies

MichaelHeerklotz opened the pull request:
See https://github.com/piwik/plugin-QueuedTracking/issues/51

MichaelHeerklotz commented:
Related PRs:
https://github.com/piwik/piwik-php-tracker/pull/28
https://github.com/piwik/piwik/pull/11771

tsteur commented on Tracker/Handler.php:
```diff
@@ -42,6 +42,11 @@ public function process(Tracker $tracker, RequestSet $requestSet)
 
+        $requests = $requestSet->getRequests();
+        foreach ($requests as $request) {
+            $request->setThirdPartyCookie($request->getVisitorId());
```
This looks good to me. FYI: I reckon this won't work as expected for a bulk request with different visitor ids.

If there were multiple requests through a bulk request, and the bulk request contains requests for visitorIds, then it would likely always overwrite the same cookie `name=0` with the last request of the request set in `setThirdPartyCookie` where it executes `$cookie->set(0, bin2hex($idVisitor));`.

Bulk requests with different visitorIds should be only the case for the log importer or when sending maybe lots of requests through the Piwik PHP tracker (or Java tracker, Android, iOS, ...) but in these cases the third party cookie wouldn't be used anyway I presume. 

So for the regular JS tracker this shouldn't be a problem and neither for other cases.

MichaelHeerklotz replied:
Interesting point. What about the other direction, when server logs are imported with the log importert tool. I took a quick look at its source code, and could not find any code that forwards the third party cookie from the logs to the tracker, so I guess it is not supported right now. Does this mean, that in this case we lose the whole third party functionality, and basically "corrupt" the data by using the log importer? Would it maybe be a good idea to create a patch for the log importer, so that it also reads the third party cookie from the logs, and only creates bulk requests for log entries with the same third party cookie?

MichaelHeerklotz replied:
In the case where the browser does not already have a third party cookie and Piwik is turned off, would it maybe be a good idea, to just generate a random new visitor id and send it to the browser? Or would this cookie then be rejected by Piwik later on, because the visitor is unknown?

tsteur replied:
@mattab I think you can answer this better. I think eg in the apache logs there is no information about cookies so they can't be replayed / applied. 

>  would it maybe be a good idea, to just generate a random new visitor id and send it to the browser

not quite sure what you mean by Piwik is turned off?

MichaelHeerklotz replied:
nginx for example can log cookies ( if I did not misunderstand the documentation). With "turned off" I mean the case when Piwik does not track anything and all requests go to the webserver log (apache/ngingx).

tsteur replied:
I don't really know what ngnix can log but I would presume it can. And then it should be possible to forward the third party cookie when importing logs. Not sure if there would be any advantage unless the cookie really already exists in Piwik. I presume it would be actually useful eg during a (short) maintenance mode where maybe many users would already have a cookie, and then this cookie could be considered but it still wouldn't be sent back to the user. And possible the idVisitor would already exist in the tracking URL anyway? Not sure! @mattab ?

When Piwik is turned off, it wouldn't be possible to send the visitor cookie I suppose, unless I misunderstand something.

tsteur commented:
Looks good to merge but we basically need to wait for https://github.com/piwik/piwik-php-tracker/pull/28 to be merged first.

MichaelHeerklotz commented:
I have updated the PR.
Just a  few cosmestic changes:
- Removed an unnecessary space
- Renamed (enable/disable)ThirdPartyCookies() to  (enable/disable)ThirdPartyCookie()

mattab commented:
> Interesting point. What about the other direction, when server logs are imported with the log importert tool. I took a quick look at its source code, and could not find any code that forwards the third party cookie from the logs to the tracker, so I guess it is not supported right now. 

Correct

> Does this mean, that in this case we lose the whole third party functionality, and basically "corrupt" the data by using the log importer?

Yes, the "third party cookie data" will be missing (ie. 'corrupt')

> Would it maybe be a good idea to create a patch for the log importer, so that it also reads the third party cookie from the logs, and only creates bulk requests for log entries with the same third party cookie?

Yes, a patch for log importer would ideally be needed. The log analytics tool would read the "third party cookie" in each log line. When a 3rd party cookie called `cookie_name = _pk_uid` is found in the line, it would be passed to the Tracking API as `_id` parameter (could also use `cid` but `cid` is prioritary over `_id` so might be better to use the lower priority in case tracking API user is using `cid` for another more important reason). 

Not sure what you meant by " and only creates bulk requests for log entries with the same third party cookie?" ?

> Of course even then, we can not "write" the third party cookie to the browser when Piwik is disabled, and would fix only the cases where the browser already received the third party cookie before Piwik was turned off.

That's a good point. When Piwik is "off" (in maintenance mode), unfortunately we would not be able to set the third party cookie, because setting the cookie requires access to the visit and database. The code is:
`$this->request->setThirdPartyCookie($this->visitProperties->getProperty('idvisitor'));`

Therefore as you point out, Replaying logs with 3rd party cookies would only work 100% and forward the 3rd party cookie for those logs created while Piwik was tracking and not in maintenance mode. But even when Piwik is in maintenance mode, we would still be able to replay those 3rd party cookies from returning visitors who had the 3rd party cookie created while Piwik was not in maintenance mode. (only new visitors coming during maintenance period would not have the 3rd party created.)

> In the case where the browser does not already have a third party cookie and Piwik is turned off, would it maybe be a good idea, to just generate a random new visitor id and send it to the browser? Or would this cookie then be rejected by Piwik later on, because the visitor is unknown?

good question. During Maintenance mode, we could indeed set the third party cookie to a random generated value. When replaying the data then, the 3rd party cookie value would be read (assuming we patch the log analytics as discussed above) and forwarded to tracking api as `_id`. So Piwik Tracking API should use this `_id` value as the visitor id when creating a new visit, and matching/searching previous visits. I think it would work. But it's a bit complicated and will need some testing to be sure.

What do you think?

mattab commented:
Note: If we patch the Log Importer to add 3rd party cookie support, would be great to also add a note in the README of log analytics to mention how to add the Cookie information  to the access logs (nginx and apache). They both support it but log format will need to be configured to include the cookie.

mattab commented:
Note: adding cookies to the log would result in huge log files and may be impractical in some cases. As I don't know / don't think it's possible to "Only log" third party cookies, also all 1st party cookies would be logged, and it would be quite big in some cases.

tsteur commented:
@MichaelHeerklotz I'd be ok to still merge as it doesn't break anything and adds support to regular JS requests. I presume making this work for eg log importer will be a different topic that would need to be discussed somewhere else. OK to merge?

MichaelHeerklotz commented:
@mattab With nginx it is possible to log only the third party cookie

MichaelHeerklotz commented:
@tsteur Yes, merging this PR would be great.
@mattab  @tsteur  Thank for your feedback about the log importer. I will try to create a PR for that next week.

tsteur commented:
Merged 👍
